File: src/lib/chunker.ts
import { config } from "./config";
import type { ParsedPage } from "./pdf-parser";
import type { ChunkMetadata } from "./types";

export interface TextChunk {
  content: string;
  pageNumber: number;
  chunkIndex: number;
  metadata: ChunkMetadata;
}

const SENTENCE_ENDINGS = /(?<=[.!?])\s+/;
const PARAGRAPH_BREAK = /\n\s*\n/;

function splitIntoSentences(text: string): string[] {
  return text
    .split(SENTENCE_ENDINGS)
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

function splitIntoParagraphs(text: string): string[] {
  return text
    .split(PARAGRAPH_BREAK)
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
}

export function createChunks(
  pages: ParsedPage[],
  chunkSize: number = config.chunking.chunkSize,
  chunkOverlap: number = config.chunking.chunkOverlap
): TextChunk[] {
  const chunks: TextChunk[] = [];
  let chunkIndex = 0;

  // Build a flat list of paragraphs tagged with their page number
  const taggedParagraphs: { text: string; pageNumber: number }[] = [];
  for (const page of pages) {
    const paragraphs = splitIntoParagraphs(page.text);
    for (const para of paragraphs) {
      taggedParagraphs.push({ text: para, pageNumber: page.pageNumber });
    }
  }

  let currentChunkParts: string[] = [];
  let currentLength = 0;
  let currentStartPage = taggedParagraphs[0]?.pageNumber ?? 1;
  let currentEndPage = currentStartPage;

  for (const { text, pageNumber } of taggedParagraphs) {
    // If a single paragraph exceeds chunk size, split it by sentences
    if (text.length > chunkSize) {
      // Flush current buffer first
      if (currentChunkParts.length > 0) {
        const content = currentChunkParts.join("\n\n");
        chunks.push(
          buildChunk(content, currentStartPage, currentEndPage, chunkIndex++)
        );
        // Keep overlap
        const overlapParts = getOverlapParts(currentChunkParts, chunkOverlap);
        currentChunkParts = overlapParts;
        currentLength = overlapParts.join("\n\n").length;
        currentStartPage = currentEndPage;
      }

      // Split large paragraph into sentence-level chunks
      const sentences = splitIntoSentences(text);
      let sentenceBuffer: string[] = [];
      let sentenceLength = 0;

      for (const sentence of sentences) {
        if (sentenceLength + sentence.length > chunkSize && sentenceBuffer.length > 0) {
          const content = sentenceBuffer.join(" ");
          chunks.push(buildChunk(content, pageNumber, pageNumber, chunkIndex++));
          // Keep overlap from sentences
          const overlapSentences = getOverlapParts(sentenceBuffer, chunkOverlap);
          sentenceBuffer = overlapSentences;
          sentenceLength = overlapSentences.join(" ").length;
        }
        sentenceBuffer.push(sentence);
        sentenceLength += sentence.length + 1;
      }

      // Remaining sentences become part of the next chunk
      if (sentenceBuffer.length > 0) {
        currentChunkParts = [sentenceBuffer.join(" ")];
        currentLength = currentChunkParts[0].length;
        currentStartPage = pageNumber;
        currentEndPage = pageNumber;
      }
      continue;
    }

    // Normal case: accumulate paragraphs
    if (currentLength + text.length + 2 > chunkSize && currentChunkParts.length > 0) {
      const content = currentChunkParts.join("\n\n");
      chunks.push(
        buildChunk(content, currentStartPage, currentEndPage, chunkIndex++)
      );
      // Keep overlap
      const overlapParts = getOverlapParts(currentChunkParts, chunkOverlap);
      currentChunkParts = overlapParts;
      currentLength = overlapParts.join("\n\n").length;
      currentStartPage = currentEndPage;
    }

    currentChunkParts.push(text);
    currentLength += text.length + 2;
    currentEndPage = pageNumber;
  }

  // Flush remaining
  if (currentChunkParts.length > 0) {
    const content = currentChunkParts.join("\n\n");
    chunks.push(
      buildChunk(content, currentStartPage, currentEndPage, chunkIndex++)
    );
  }

  return chunks;
}

function buildChunk(
  content: string,
  startPage: number,
  endPage: number,
  chunkIndex: number
): TextChunk {
  const words = content.split(/\s+/).filter((w) => w.length > 0);
  return {
    content,
    pageNumber: startPage,
    chunkIndex,
    metadata: {
      char_count: content.length,
      word_count: words.length,
      start_page: startPage,
      end_page: endPage,
    },
  };
}

function getOverlapParts(parts: string[], overlapSize: number): string[] {
  const result: string[] = [];
  let totalLength = 0;

  for (let i = parts.length - 1; i >= 0; i--) {
    if (totalLength >= overlapSize) break;
    result.unshift(parts[i]);
    totalLength += parts[i].length;
  }

  return result;
}

File: internal/chunker/chunker.go
package chunker

import (
	"regexp"
	"strings"

	"github.com/docinsight/backend/internal/model"
	"github.com/docinsight/backend/internal/pdf"
	"github.com/google/uuid"
)

var (
	// Go's RE2 regexp doesn't support lookbehinds, so we match the punctuation
	// plus trailing whitespace and re-append the punctuation in splitIntoSentences.
	sentenceEndings = regexp.MustCompile(`([.!?])\s+`)
	paragraphBreak  = regexp.MustCompile(`\n\s*\n`)
)

type TextChunk struct {
	Content    string
	PageNumber int
	ChunkIndex int
	Metadata   model.ChunkMetadata
}

type taggedParagraph struct {
	text       string
	pageNumber int
}

func CreateChunks(pages []pdf.Page, chunkSize, chunkOverlap int) []TextChunk {
	if len(pages) == 0 {
		return nil
	}

	var tagged []taggedParagraph
	for _, page := range pages {
		paragraphs := splitIntoParagraphs(page.Text)
		for _, p := range paragraphs {
			tagged = append(tagged, taggedParagraph{text: p, pageNumber: page.Number})
		}
	}

	if len(tagged) == 0 {
		return nil
	}

	var chunks []TextChunk
	chunkIndex := 0

	var currentParts []string
	currentLength := 0
	currentStartPage := tagged[0].pageNumber
	currentEndPage := currentStartPage

	for _, tp := range tagged {
		// If a single paragraph exceeds chunk size, split by sentences
		if len(tp.text) > chunkSize {
			// Flush current buffer
			if len(currentParts) > 0 {
				content := strings.Join(currentParts, "\n\n")
				chunks = append(chunks, buildChunk(content, currentStartPage, currentEndPage, chunkIndex))
				chunkIndex++

				overlapParts := getOverlapParts(currentParts, chunkOverlap)
				currentParts = overlapParts
				currentLength = joinLength(overlapParts, "\n\n")
				currentStartPage = currentEndPage
			}

			// Split large paragraph into sentence-level chunks
			sentences := splitIntoSentences(tp.text)
			var sentBuf []string
			sentLen := 0

			for _, sent := range sentences {
				if sentLen+len(sent) > chunkSize && len(sentBuf) > 0 {
					content := strings.Join(sentBuf, " ")
					chunks = append(chunks, buildChunk(content, tp.pageNumber, tp.pageNumber, chunkIndex))
					chunkIndex++

					overlapSentences := getOverlapParts(sentBuf, chunkOverlap)
					sentBuf = overlapSentences
					sentLen = joinLength(overlapSentences, " ")
				}
				sentBuf = append(sentBuf, sent)
				sentLen += len(sent) + 1
			}

			if len(sentBuf) > 0 {
				currentParts = []string{strings.Join(sentBuf, " ")}
				currentLength = len(currentParts[0])
				currentStartPage = tp.pageNumber
				currentEndPage = tp.pageNumber
			}
			continue
		}

		// Normal case: accumulate paragraphs
		if currentLength+len(tp.text)+2 > chunkSize && len(currentParts) > 0 {
			content := strings.Join(currentParts, "\n\n")
			chunks = append(chunks, buildChunk(content, currentStartPage, currentEndPage, chunkIndex))
			chunkIndex++

			overlapParts := getOverlapParts(currentParts, chunkOverlap)
			currentParts = overlapParts
			currentLength = joinLength(overlapParts, "\n\n")
			currentStartPage = currentEndPage
		}

		currentParts = append(currentParts, tp.text)
		currentLength += len(tp.text) + 2
		currentEndPage = tp.pageNumber
	}

	// Flush remaining
	if len(currentParts) > 0 {
		content := strings.Join(currentParts, "\n\n")
		chunks = append(chunks, buildChunk(content, currentStartPage, currentEndPage, chunkIndex))
	}

	return chunks
}

func ToModelChunks(textChunks []TextChunk, documentID uuid.UUID) []model.Chunk {
	chunks := make([]model.Chunk, len(textChunks))
	for i, tc := range textChunks {
		chunks[i] = model.Chunk{
			ID:         uuid.New(),
			DocumentID: documentID,
			Content:    tc.Content,
			PageNumber: tc.PageNumber,
			ChunkIndex: tc.ChunkIndex,
			Metadata:   tc.Metadata,
		}
	}
	return chunks
}

func buildChunk(content string, startPage, endPage, chunkIndex int) TextChunk {
	words := strings.Fields(content)
	return TextChunk{
		Content:    content,
		PageNumber: startPage,
		ChunkIndex: chunkIndex,
		Metadata: model.ChunkMetadata{
			CharCount: len(content),
			WordCount: len(words),
			StartPage: startPage,
			EndPage:   endPage,
		},
	}
}

func splitIntoParagraphs(text string) []string {
	parts := paragraphBreak.Split(text, -1)
	var result []string
	for _, p := range parts {
		trimmed := strings.TrimSpace(p)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

func splitIntoSentences(text string) []string {
	// Find all sentence-ending punctuation + whitespace boundaries
	matches := sentenceEndings.FindAllStringIndex(text, -1)
	if len(matches) == 0 {
		trimmed := strings.TrimSpace(text)
		if trimmed == "" {
			return nil
		}
		return []string{trimmed}
	}

	var result []string
	prev := 0
	for _, m := range matches {
		// Include the punctuation character (m[0] is start of "[.!?]\s+")
		// The sentence runs from prev to m[0]+1 (include the punctuation char)
		sentence := strings.TrimSpace(text[prev : m[0]+1])
		if sentence != "" {
			result = append(result, sentence)
		}
		prev = m[1] // after the whitespace
	}
	// Remaining text after last match
	if prev < len(text) {
		trimmed := strings.TrimSpace(text[prev:])
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

func getOverlapParts(parts []string, overlapSize int) []string {
	var result []string
	totalLen := 0

	for i := len(parts) - 1; i >= 0; i-- {
		if totalLen >= overlapSize {
			break
		}
		result = append([]string{parts[i]}, result...)
		totalLen += len(parts[i])
	}

	return result
}

func joinLength(parts []string, sep string) int {
	if len(parts) == 0 {
		return 0
	}
	total := 0
	for _, p := range parts {
		total += len(p)
	}
	total += len(sep) * (len(parts) - 1)
	return total
}
